### PR TITLE
Add basic DTS

### LINKED ISSUE #7: TypeScript Conversion
This looks really great. How would you feel about converting to TypeScript? (happy to help)

### PR DESCRIPTION
**Pull request by pspeter3**
This DTS file does not support the JSX namespace because that is a more complicated issue.

Closes #7

### CODE REVIEW

**Comment by ngryman** [comment]
In #7 you speak about tests, do you still want to add them?

**Comment by pspeter3** [comment]
Given the DTS size, I think that you're right and tests are not needed at this point.

**Comment by jbucaran** [comment]
> @pspeter3 One issue though is it seems like you can not have a component that returns another component when using h. Thoughts?

Do you mean?

```jsx
const AnotherComponent = (props, children) => h("div", {}, "hi")
```
```jsx
const Component = (props, children) => AnotherComponent
```

or

```jsx
const Component = (props, children) => AnotherComponent(props, children)
```

**Comment by pspeter3** [comment]
I mean the middle one where the component returns another function. I'm not sure a world which that would happen and the implementation of `h` does not handle it so I didn't make the DTS handle it. I'm not particularly worried about it.

**Comment by jbucaran** [comment]
@pspeter3 Got it. Note that the following does work:

```jsx
const Component = (props, children) => h(AnotherComponent, props, children)
```

**Comment by pspeter3** [comment]
@jbucaran do we still want this?

**Comment by jbucaran** [comment]
@pspeter3 Definitely! 

For reference: https://github.com/hyperapp/hyperapp/pull/279

**Comment by pspeter3** [comment]
So what's next to have this merged?

File: package.json [hunk 1]
```diff
@@ -19,11 +19,12 @@
...
-    "format": "prettier --semi false --write *.js"
+    "format": "prettier --semi false --write *.js *.d.ts"
   },
   "devDependencies": {
     "prettier": "~1.4.2",
     "rollup": "^0.42.0",
     "rollup-plugin-uglify": "^2.0.1"
-  }
+  },
+  "typings": "picodom.d.ts"
```
Comment (jbucaran) [inline package.json]: I thought this was outdated? 

/cc @Swizz

**Comment by jbucaran** [comment]
@pspeter3 Please fix the conflict first! 🙏👍

File: package.json (hunk 1 above)
Comment (Swizz) [inline package.json]: No, not this one, the one you requested about the `typings.json` file

**Comment by pspeter3** [comment]
Fixed

**Comment by pspeter3** [comment]
Ping?

File: picodom.d.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    tag: string
+    data?: D
+    children: VNode<{}>
+  }
+
+  export interface Component<D> {
+    (data?: D, ...children: VNode<{}>[]): VNode<D>
+  }
+
+  export type VElement<D> = Component<D> | string
```
Comment (jbucaran) [inline picodom.d.ts]: @pspeter3 Can we somehow avoid declaring VElement and just put it in the definition of `tag` below?

File: picodom.d.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  export function h<D>(
+    tag: VElement<D>,
+    data?: D,
+    ...children: VNode<{}>[]
```
Comment (jbucaran) [inline picodom.d.ts]: @pspeter3 Does this mean variable Vnode args or array of Vnodes?

File: picodom.d.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  ): VNode<D>
+
+  export function patch(
+    parent: Element,
```
Comment (jbucaran) [inline picodom.d.ts]: @pspeter3 Is Element the same as HTMLElement?

File: picodom.d.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    element: Element,
```
Comment (jbucaran) [inline picodom.d.ts]: element can be null.

File: picodom.d.ts (hunk 4 above)
Comment (andrewiggins) [inline picodom.d.ts]: Element is base type for HTMLElement and SVGElement

File: picodom.d.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+declare namespace picodom {
+  export interface VNode<D> {
+    tag: string
+    data?: D
+    children: VNode<{}>
+  }
+
+  export interface Component<D> {
+    (data?: D, ...children: VNode<{}>[]): VNode<D>
```
Comment (andrewiggins) [inline picodom.d.ts]: Does this map to the last line in h.js: `tag(data, children)`? If so, should this be 
```typescript
export type VNodeChildren = VNode<{}> | string;

export interface Component<D> {
    (data?: D, children: VNodeChildren[]): VNode<D>
}
```

Bonus, you can reuse `VNodeChildren` in the `VNode` interface.

File: picodom.d.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+declare namespace picodom {
```
Comment (andrewiggins) [inline picodom.d.ts]: Thoughts on we following the [module.d.ts UMD example](http://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html) from the [TypeScript handbook](http://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#umd)?

Basically would just be removing the two declares and adding `export as namespace picodom`;

File: picodom.d.ts (hunk 6 above)
Comment (andrewiggins) [inline picodom.d.ts]: Similar to [Optional Parameters in Callbacks](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#optional-parameters-in-callbacks), I don't think data needs to be optional here. The Component interface will always be invoked with a `data` argument and `children` argument. If a consumer wants to ignore them that is valid.

I tested the following locally and it passes:
```typescript
export interface Component<D> {
  (data: D, children: VNode<{}>[]): VNode<D>
}

let Test: Component<any> = () => <div></div>;
```

File: package.json [hunk 8]
```diff
@@ -35,5 +35,6 @@
     "prettier": "~1.5.3",
     "rollup": "^0.45.2",
     "rollup-plugin-uglify": "^2.0.1"
-  }
+  },
+  "typings": "picodom.d.ts"
```
Comment (andrewiggins) [inline package.json]: Should this be types ([doc](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html))?

**Comment by andrewiggins** [comment]
For the comments I left, I'd be happy to jump in a contribute some fixes. I'd also be happy to add some JSDoc comments to the definitions for better intellisense in editors such as VSCode

**Comment by andrewiggins** [comment]
At some point (either this PR or another) we should look at exposing the available lifecycle methods through the type definitions

File: picodom.d.ts (hunk 6 above)
Comment (jbucaran) [inline picodom.d.ts]: This is correct, data is a required argument. It can be `null` or `{}` if you want to ignore it.

Comment (jbucaran) [inline picodom.d.ts]: @pspeter3 By the way, I'd prefer to rename this argument to `props` instead.

**Comment by jbucaran** [comment]
> @andrewiggins At some point (either this PR or another) we should look at exposing the available lifecycle methods through the type definitions.

Sounds good to me.

File: picodom.d.ts (hunk 2 above)
Comment (pspeter3) [inline picodom.d.ts]: Yes but why? It's been helpful with the React type definitions to have this type exist. I actually think it needs to be applied to more places.

File: picodom.d.ts (hunk 7 above)
Comment (pspeter3) [inline picodom.d.ts]: That works, I haven't done it in a while. Will update to that.

File: picodom.d.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+export interface VNode<D> {
+  tag: string
+  data?: D
+  children: VElement<{}>[]
```
Comment (jbucaran) [inline picodom.d.ts]: @pspeter3 I think we might not need VElement, but please correct me if I am wrong. The signature of a Component is good, but a virtual node children property is always an array. We make sure of that inside `h`.

Comment (pspeter3) [inline picodom.d.ts]: Yeah it's an array of `VNode | string` right? How else do you represent text nodes?

Comment (jbucaran) [inline picodom.d.ts]: See what I mean:

<pre>
h("div", {}, "Foo") => { 
  tag: "div", 
  data: {}, 
  children: <b>["Foo"]</b> 
}
</pre>

Comment (jbucaran) [inline picodom.d.ts]: Children is always an array. If you pass a string to h(), it will return Array<string>.

Comment (pspeter3) [inline picodom.d.ts]: So VElement just means `VNode | string` which is what you have there. It's an array of them which is also what you have there.

**Comment by jbucaran** [comment]
@pspeter3 Is this merge ready? and how can I test if this is working, like, how do I tell vscode to pick the typings? 

Sorry, total typings n00b here.

**Comment by pspeter3** [comment]
Yeah, I think we can merge it.

**Comment by jbucaran** [comment]
@pspeter3 Any tips on how to tell vscode to pick this up?

File: picodom.d.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+export function h<D>(
+  tag: Component<D> | string,
+  data?: D,
+  ...children: VElement<{}>[]
+): VNode<D>
+
+export function patch(
+  parent: Element,
+  element: Element,
```
Comment (jbucaran) [inline picodom.d.ts]: @pspeter3 element can be null.

File: picodom.d.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+export interface Component<D> {
+  (data?: D, ...children: VElement<{}>[]): VNode<D>
+}
+
+export type VElement<D> = VNode<D> | string
+
+export function h<D>(
+  tag: Component<D> | string,
+  data?: D,
```
Comment (jbucaran) [inline picodom.d.ts]: Can we rename `D` to something more meaningful?

- Data
- DataAttributes
- VNodeData
- VNodeAttributes

Comment (andrewiggins) [inline picodom.d.ts]: * Props?

Comment (pspeter3) [inline picodom.d.ts]: I think Data makes sense since I took it from the function signature.

File: picodom.d.ts (hunk 10 above)
Comment (pspeter3) [inline picodom.d.ts]: Oh, good to know.

File: picodom.d.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
+export interface VNode<Data> {
+  tag: string
+  data?: Data
+  children: VElement<{}>[]
+}
+
+export interface Component<Data> {
+  (data?: D, ...children: VElement<{}>[]): VNode<D>
```
Comment (jbucaran) [inline picodom.d.ts]: D

Comment (pspeter3) [inline picodom.d.ts]: I wish you had the ability to edit my PR. Good catch though

**Comment by jbucaran** [comment]
Thank you @pspeter3 for your incredible patience with this! 🎉

**Comment by Swizz** [comment]
I am coming too late, **but** I am reworking the Hyperapp DTS largely based on the picodom DTS (because you did on awesome job !)

So, `h()` also accept as parameter a children that will be an array of Vnode, so, why not overloads `h()` with : 
```tsx
export function h<Data>(
  tag: Component<Data> | string,
  data?: Data,
  children?: VElement<{}>[]
): VNode<Data>
```
Plus, `h()` is ready to be use with JSX, so why not type JSX support too ?

```tsx
declare global {
  namespace JSX {
    interface Element<Data> extends VNode<Data> {}
    interface IntrinsicElements {
      [elemName: string]: {}
    }
  }
}
```

I am sorry if this comment could be judged as pretty annoying.

**Comment by andrewiggins** [comment]
@Swizz I'm working on improving the types a bit, including what you mentioned. PR should be out soon.

**Pull request merged**